{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np, pandas as pd\n",
    "import re\n",
    "import twobitreader\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The cells below detail how to annotate a call stats file to be used as a training set\n",
    "\n",
    "## Before running, make sure that the call stats are annotated for overlaps with segmental duplications, germline copy number gains and have UCSC mappability scores. Filtering out overlap with Hardy Weinberg failures also recommended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "File b'all_clone/blacklist/exac_cluster_allClone_hwe_segdup_CNV.maf' does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-3a2489760cda>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0moutput_filename\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"all_clone/blacklist/exac_cluster_allClone_hwe_segdup_CNV_proxy.maf\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mtrain\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_filename\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mtrain\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtrain\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'contig'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"position\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    676\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    677\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 678\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    679\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    680\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    439\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 440\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    441\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    785\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    786\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 787\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    788\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    789\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1012\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1013\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1014\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1015\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1016\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1706\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'usecols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1707\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1708\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1709\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1710\u001b[0m         \u001b[0mpassed_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnames\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: File b'all_clone/blacklist/exac_cluster_allClone_hwe_segdup_CNV.maf' does not exist"
     ]
    }
   ],
   "source": [
    "input_filename=\"cs.txt\"\n",
    "output_filename=\"cs_af_next.txt\"\n",
    "\n",
    "train=pd.read_csv(input_filename,sep='\\t')\n",
    "\n",
    "train=train.sort_values(by=['contig',\"position\"])\n",
    "\n",
    "train=train.reset_index(drop=True)\n",
    "\n",
    "train['af_next']=train['tumor_f']\n",
    "train['af_prev']=train['tumor_f']\n",
    "\n",
    "#get the allele fractions of the next SNP\n",
    "train['af_next']=train['af_next'].shift(periods=-1)\n",
    "\n",
    "#get allele fractions of the previous SNP\n",
    "train['af_prev']=train['af_prev'].shift()\n",
    "\n",
    "train.to_csv(output_filename,sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename alphabetical chromosomes to numbers\n",
    "def chrom2int(x):\n",
    "    if x == 'X':\n",
    "        return 23\n",
    "    elif x == 'Y':\n",
    "        return 24\n",
    "    else:\n",
    "        return int(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66893\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mleventh/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:42: FutureWarning: 'genomic_position' is both an index level and a column label.\n",
      "Defaulting to column, but this will raise an ambiguity error in a future version\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34225\n",
      "['1' '2' '3' '4' '5' '6' '7' '8' '9' '10' '11' '12' '13' '14' '15' '16'\n",
      " '17' '18' '19' '20' '21' '22' 'X' 'Y']\n",
      "[21504264.0, 21504264.0]\n",
      "34225\n"
     ]
    }
   ],
   "source": [
    "## Annotating training file for the distribution of allele fractions at a given site and the distance to the\n",
    "## nearest SNP\n",
    "input_filename=\"cs_af_next.txt\"\n",
    "train=pd.read_csv(input_filename,sep='\\t')\n",
    "train['contig']=train['contig'].apply(str)\n",
    "\n",
    "pattern=r'^[0-9]+$'\n",
    "pattern_X=r'^X$'\n",
    "pattern_Y=r'^Y$'\n",
    "\n",
    "train=train.loc[(train[\"contig\"].str.contains(pattern))|(train[\"contig\"].str.contains(pattern_X))|\n",
    "                (train[\"contig\"].str.contains(pattern_Y)),:]\n",
    "train.reset_index(drop=True,inplace=True)\n",
    "print(len(train.index))\n",
    "\n",
    "\n",
    "chr_lens = [0, 249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,\n",
    "              146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,\n",
    "              102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,\n",
    "           155270560]\n",
    "\n",
    "offsets = np.cumsum(chr_lens)\n",
    "\n",
    "chrom=[]\n",
    "\n",
    "for idx,row in train.iterrows():\n",
    "    chrom.append(chrom2int(row.contig))\n",
    "\n",
    "train[\"Chromosome\"]=chrom\n",
    "train[\"genomic_position\"]=pd.to_numeric(train[\"position\"])+offsets[\n",
    "    pd.to_numeric(train[\"Chromosome\"])-1\n",
    "]\n",
    "\n",
    "train['af_category']=train['tumor_f']//0.1\n",
    "\n",
    "cat_counts=train.groupby([\"genomic_position\",\"af_category\"]).size().reset_index(name='counts')\n",
    "cat_counts=cat_counts.pivot(index='genomic_position',columns='af_category',values='counts')\n",
    "cat_counts=cat_counts.fillna(0)\n",
    "\n",
    "cat_counts['row_sum']=cat_counts.iloc[:,0:len(cat_counts.columns)].sum(axis=1)\n",
    "cat_counts=cat_counts.iloc[:,0:len(cat_counts.columns)-1].div(cat_counts.loc[:,'row_sum'],axis=0)\n",
    "cat_counts['genomic_position']=cat_counts.index\n",
    "\n",
    "train=train.merge(cat_counts,left_on=\"genomic_position\",right_on=\"genomic_position\")\n",
    "\n",
    "\n",
    "train.rename(columns={0.0: \"less_than_0.1\", 1.0: \"between_0.1_0.2\",\n",
    "                                     2.0:\"between_0.2_0.3\",3.0:\"between_0.3_0.4\",\n",
    "                                     4.0:\"between_0.4_0.5\",5.0:\"between_0.5_0.6\"},inplace=True)\n",
    "\n",
    "mean_val=train.groupby([\"genomic_position\"])['tumor_f'].mean().reset_index(name='tumor_f')\n",
    "print(len(mean_val.index))\n",
    "\n",
    "train=mean_val.merge(train,left_on=\"genomic_position\",right_on=\"genomic_position\")\n",
    "train=train.reset_index(drop=True)\n",
    "\n",
    "train=train.drop_duplicates(subset='genomic_position', keep='first')\n",
    "\n",
    "train=train.sort_values(by=[\"genomic_position\"])\n",
    "\n",
    "train['distance_prev']=train['genomic_position'].diff()\n",
    "train['distance_next']=abs(train['genomic_position'].diff(periods=-1))\n",
    "print(train.contig.unique())\n",
    "\n",
    "print([np.max(train['distance_prev']),np.max(train['distance_next'])])\n",
    "print(len(train.index))\n",
    "\n",
    "train.to_csv(\"small_case_shuffled_HWE_segdup_commonCN_map_proxy_dist.txt\",sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The following cells can be used to either annotate a training set or annotate a somatic mutation calling set for SVM prediciton. \n",
    "\n",
    "## Like the call stats file, the somatic mutation calling calling file should be annotated for overlaps with segmental duplications, germline CNVs and UCSC mappability scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mleventh/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "## annotate for proximity to bait boundary\n",
    "train=pd.read_csv(\"test.txt\",sep='\\t')\n",
    "agilent=pd.read_csv(\"whole_exome_agilent_1.1_refseq_plus_3_boosters.Homo_sapiens_assembly19.targets.bed\",sep='\\t')\n",
    "\n",
    "chrom=[]\n",
    "\n",
    "for idx,row in train.iterrows():\n",
    "    #run if using a call stats file\n",
    "    #chrom.append(chrom2int(row.contig))\n",
    "    \n",
    "    #run if using an annotated maf\n",
    "    chrom.append(chrom2int(row.Chromosome))\n",
    "\n",
    "train[\"Chromosome_num\"]=chrom\n",
    "\n",
    "chrom=[]\n",
    "for idx,row in agilent.iterrows():\n",
    "    chrom.append(chrom2int(row.Chromosome))\n",
    "\n",
    "agilent[\"Chromosome\"]=chrom\n",
    "\n",
    "agilent=agilent.dropna(subset=[\"Chromosome\",\"Start_position\",\"End_position\"],axis=0)\n",
    "bait_boundary=[]\n",
    "for idx,row in train.iterrows():\n",
    "    \n",
    "    agilent_chrom=agilent.loc[agilent['Chromosome']==row.Chromosome_num,:]\n",
    "\n",
    "    #for call stats\n",
    "    #diff_start=abs(agilent_chrom.Start_position-row.position)\n",
    "    #diff_end=abs(agilent_chrom.End_position-row.position)\n",
    "    \n",
    "    #for annotated maf\n",
    "    diff_start=abs(agilent_chrom.Start_position-row.Start_position)\n",
    "    diff_end=abs(agilent_chrom.End_position-row.Start_position)\n",
    "    \n",
    "    boundary=min(min(diff_start),min(diff_end))\n",
    "\n",
    "    bait_boundary.append(boundary)\n",
    "    \n",
    "train['distance_to_bait']=bait_boundary\n",
    "\n",
    "train.to_csv(\"test_bait.txt\",sep='\\t',index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## annotate GC content of 75mer context, can be slow\n",
    "def chrom2int(x):\n",
    "    if x == 'X':\n",
    "        return 23\n",
    "    elif x == 'Y':\n",
    "        return 24\n",
    "    else:\n",
    "        return int(x)\n",
    "\n",
    "def get_reference(genome,chrom,start,end):\n",
    "\n",
    "    contig=genome[str(chrom)]\n",
    "    return contig[start:end+1]\n",
    "\n",
    "def find_reads(chrm,pos,rep_len):\n",
    "\n",
    "    # initialize 2bitreader genome\n",
    "    genome = twobitreader.TwoBitFile(\"hg19.2bit\")\n",
    "    chr=[genome[c] for c in map(str,range(1,23))]\n",
    "    chr.append(genome['X'])\n",
    "    chr.append(genome['Y'])\n",
    "\n",
    "    read=get_reference(genome,chrm,pos-(rep_len//2)-1,pos+(rep_len//2)-1)\n",
    "    \n",
    "    return read\n",
    "\n",
    "def gc_content(read):\n",
    "    return (read.count('G')+read.count('C'))/len(read)\n",
    "\n",
    "train=pd.read_csv(\"test_bait.txt\",sep='\\t')\n",
    "train['gc_content']=0\n",
    "\n",
    "for idx,row in train.iterrows():\n",
    "    #gc_read=find_reads(row['contig'],row['position'],75)\n",
    "    gc_read=find_reads(row['Chromosome'],row['Start_position'],75)\n",
    "    train.loc[idx,'gc_content']=gc_content(gc_read)\n",
    "\n",
    "train.to_csv(\"test_gc.txt\",\n",
    "             sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "ANNOTATE SOMATIC MUTATIONS WITH NEAREST GERMLINE AF, POSITION, AND RECURRENCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['tumor_f', 'ref_allele', 'context', 'contig', 'position', 'hwe_fail',\n",
      "       'has_segdup', 'gain_overlap', 'loss_overlap', 'score', 'af_next',\n",
      "       'af_prev', 'less_than_0.1', 'between_0.1_0.2', 'between_0.2_0.3',\n",
      "       'between_0.3_0.4', 'between_0.4_0.5', 'between_0.5_0.6',\n",
      "       'distance_prev', 'distance_next', 'Chromosome', 'distance_to_bait',\n",
      "       'gc_content'],\n",
      "      dtype='object')\n",
      "Index(['Hugo_Symbol', 'Entrez_Gene_Id', 'Center', 'NCBI_Build', 'Chromosome',\n",
      "       'Start_position', 'End_position', 'Strand', 'Variant_Classification',\n",
      "       'Variant_Type', 'Reference_Allele', 'Tumor_Seq_Allele1',\n",
      "       'Tumor_Seq_Allele2', 'Protein_Change', 'Cluster_ID', 'CCF_hat_01',\n",
      "       'CCF90_low_01', 'CCF90_high_01', 'Tumor_Sample_Barcode',\n",
      "       'germline_cluster', 'hwe_fail', 'gain_overlap', 'has_segdup',\n",
      "       't_alt_count', 't_ref_count', 'SwissProt_acc_Id', 'SwissProt_entry_Id',\n",
      "       'ref_context', 'pon_loglike', 'i_t_ALT_F1R2', 'i_t_ALT_F2R1',\n",
      "       'i_t_REF_F1R2', 'i_t_REF_F2R1', 't_lod_fstar', 'Chromosome_num',\n",
      "       'distance_to_bait', 'gc_content', 'score'],\n",
      "      dtype='object')\n",
      "58909\n",
      "[]\n",
      "58909\n",
      "         Hugo_Symbol  Entrez_Gene_Id         Center  NCBI_Build Chromosome_x  \\\n",
      "0             SAMD11               0  broad.mit.edu          37            1   \n",
      "1             SAMD11               0  broad.mit.edu          37            1   \n",
      "2              NOC2L               0  broad.mit.edu          37            1   \n",
      "3             KLHL17               0  broad.mit.edu          37            1   \n",
      "4             KLHL17               0  broad.mit.edu          37            1   \n",
      "5             KLHL17               0  broad.mit.edu          37            1   \n",
      "6             KLHL17               0  broad.mit.edu          37            1   \n",
      "7             KLHL17               0  broad.mit.edu          37            1   \n",
      "8            PLEKHN1               0  broad.mit.edu          37            1   \n",
      "9            PLEKHN1               0  broad.mit.edu          37            1   \n",
      "10              AGRN               0  broad.mit.edu          37            1   \n",
      "11              AGRN               0  broad.mit.edu          37            1   \n",
      "12              AGRN               0  broad.mit.edu          37            1   \n",
      "13              AGRN               0  broad.mit.edu          37            1   \n",
      "14              AGRN               0  broad.mit.edu          37            1   \n",
      "15              AGRN               0  broad.mit.edu          37            1   \n",
      "16              AGRN               0  broad.mit.edu          37            1   \n",
      "17              AGRN               0  broad.mit.edu          37            1   \n",
      "18          C1orf159               0  broad.mit.edu          37            1   \n",
      "19     RP11-465B22.8               0  broad.mit.edu          37            1   \n",
      "20            TTLL10               0  broad.mit.edu          37            1   \n",
      "21           TNFRSF4               0  broad.mit.edu          37            1   \n",
      "22           TNFRSF4               0  broad.mit.edu          37            1   \n",
      "23           TNFRSF4               0  broad.mit.edu          37            1   \n",
      "24           TNFRSF4               0  broad.mit.edu          37            1   \n",
      "25              SDF4               0  broad.mit.edu          37            1   \n",
      "26              SDF4               0  broad.mit.edu          37            1   \n",
      "27           B3GALT6               0  broad.mit.edu          37            1   \n",
      "28            SCNN1D               0  broad.mit.edu          37            1   \n",
      "29             ACAP3               0  broad.mit.edu          37            1   \n",
      "...              ...             ...            ...         ...          ...   \n",
      "44181           GDI1               0  broad.mit.edu          37            X   \n",
      "44182           GDI1               0  broad.mit.edu          37            X   \n",
      "44183         FAM50A               0  broad.mit.edu          37            X   \n",
      "44184         PLXNA3               0  broad.mit.edu          37            X   \n",
      "44185         PLXNA3               0  broad.mit.edu          37            X   \n",
      "44186         PLXNA3               0  broad.mit.edu          37            X   \n",
      "44187         PLXNA3               0  broad.mit.edu          37            X   \n",
      "44188         PLXNA3               0  broad.mit.edu          37            X   \n",
      "44189           G6PD               0  broad.mit.edu          37            X   \n",
      "44190           GAB3               0  broad.mit.edu          37            X   \n",
      "44191           GAB3               0  broad.mit.edu          37            X   \n",
      "44192           GAB3               0  broad.mit.edu          37            X   \n",
      "44193           DKC1               0  broad.mit.edu          37            X   \n",
      "44194           MPP1               0  broad.mit.edu          37            X   \n",
      "44195           MPP1               0  broad.mit.edu          37            X   \n",
      "44196             F8               0  broad.mit.edu          37            X   \n",
      "44197             F8               0  broad.mit.edu          37            X   \n",
      "44198             F8               0  broad.mit.edu          37            X   \n",
      "44199             F8               0  broad.mit.edu          37            X   \n",
      "44200             F8               0  broad.mit.edu          37            X   \n",
      "44201             F8               0  broad.mit.edu          37            X   \n",
      "44202             F8               0  broad.mit.edu          37            X   \n",
      "44203             F8               0  broad.mit.edu          37            X   \n",
      "44204             F8               0  broad.mit.edu          37            X   \n",
      "44205          MTCP1               0  broad.mit.edu          37            X   \n",
      "44206          BRCC3               0  broad.mit.edu          37            X   \n",
      "44207           VBP1               0  broad.mit.edu          37            X   \n",
      "44208            UTY               0  broad.mit.edu          37            Y   \n",
      "44209            UTY               0  broad.mit.edu          37            Y   \n",
      "44210        Unknown               0  broad.mit.edu          37            Y   \n",
      "\n",
      "       Start_position  End_position Strand Variant_Classification  \\\n",
      "0              865664        865664      +      Missense_Mutation   \n",
      "1              874501        874501      +      Missense_Mutation   \n",
      "2              887901        887901      +      Missense_Mutation   \n",
      "3              896945        896945      +                 Intron   \n",
      "4              897009        897009      +            Splice_Site   \n",
      "5              898466        898466      +                 Intron   \n",
      "6              898925        898925      +                 Intron   \n",
      "7              901964        901964      +                    IGR   \n",
      "8              906100        906100      +      Missense_Mutation   \n",
      "9              909874        909874      +                 Silent   \n",
      "10             957703        957703      +                 Silent   \n",
      "11             978753        978753      +      Missense_Mutation   \n",
      "12             979268        979268      +      Missense_Mutation   \n",
      "13             980641        980641      +                 Silent   \n",
      "14             981114        981114      +            Splice_Site   \n",
      "15             981376        981376      +      Missense_Mutation   \n",
      "16             982962        982962      +      Missense_Mutation   \n",
      "17             983000        983000      +                 Silent   \n",
      "18            1018351       1018351      +                 Silent   \n",
      "19            1104399       1104399      +                lincRNA   \n",
      "20            1114811       1114811      +                 Intron   \n",
      "21            1147001       1147001      +                 Silent   \n",
      "22            1147024       1147024      +                 Intron   \n",
      "23            1147434       1147434      +                 Silent   \n",
      "24            1148868       1148868      +                 Intron   \n",
      "25            1153084       1153084      +                 Intron   \n",
      "26            1154374       1154374      +                 Intron   \n",
      "27            1170197       1170197      +                  3'UTR   \n",
      "28            1223057       1223057      +                 Silent   \n",
      "29            1226797       1226797      +                    IGR   \n",
      "...               ...           ...    ...                    ...   \n",
      "44181       153670902     153670902      +      Missense_Mutation   \n",
      "44182       153670956     153670956      +                 Silent   \n",
      "44183       153674876     153674876      +      Missense_Mutation   \n",
      "44184       153690629     153690629      +                 Silent   \n",
      "44185       153693158     153693158      +      Missense_Mutation   \n",
      "44186       153693767     153693767      +            Splice_Site   \n",
      "44187       153697051     153697051      +      Missense_Mutation   \n",
      "44188       153698331     153698331      +      Missense_Mutation   \n",
      "44189       153762560     153762560      +      Missense_Mutation   \n",
      "44190       153940535     153940535      +                 Silent   \n",
      "44191       153940642     153940642      +      Missense_Mutation   \n",
      "44192       153940672     153940672      +      Missense_Mutation   \n",
      "44193       154003515     154003515      +                 Silent   \n",
      "44194       154007581     154007581      +                 Silent   \n",
      "44195       154020084     154020084      +                 Silent   \n",
      "44196       154133094     154133094      +      Missense_Mutation   \n",
      "44197       154133145     154133145      +      Missense_Mutation   \n",
      "44198       154134731     154134731      +                 Silent   \n",
      "44199       154157413     154157413      +      Missense_Mutation   \n",
      "44200       154157804     154157804      +      Missense_Mutation   \n",
      "44201       154159372     154159372      +      Missense_Mutation   \n",
      "44202       154185319     154185319      +                 Silent   \n",
      "44203       154194370     154194370      +      Missense_Mutation   \n",
      "44204       154197625     154197625      +                 Silent   \n",
      "44205       154293988     154293988      +      Missense_Mutation   \n",
      "44206       154317537     154317537      +            Splice_Site   \n",
      "44207       154444750     154444750      +                  5'UTR   \n",
      "44208        15447539      15447539      +      Missense_Mutation   \n",
      "44209        15591418      15591418      +      Missense_Mutation   \n",
      "44210        21749313      21749313      +                    IGR   \n",
      "\n",
      "      Variant_Type          ...          between_0.2_0.3 between_0.3_0.4  \\\n",
      "0              SNP          ...                      0.0             0.0   \n",
      "1              SNP          ...                      0.0             0.0   \n",
      "2              SNP          ...                      0.0             0.0   \n",
      "3              SNP          ...                      0.0             0.0   \n",
      "4              SNP          ...                      0.0             0.0   \n",
      "5              SNP          ...                      0.0             0.0   \n",
      "6              SNP          ...                      0.0             0.0   \n",
      "7              SNP          ...                      0.0             0.0   \n",
      "8              SNP          ...                      0.0             0.0   \n",
      "9              SNP          ...                      0.0             0.0   \n",
      "10             SNP          ...                      0.0             0.0   \n",
      "11             SNP          ...                      0.0             0.0   \n",
      "12             SNP          ...                      0.0             0.0   \n",
      "13             SNP          ...                      0.0             0.0   \n",
      "14             SNP          ...                      0.0             0.0   \n",
      "15             SNP          ...                      0.0             0.0   \n",
      "16             SNP          ...                      0.0             0.0   \n",
      "17             SNP          ...                      0.0             0.0   \n",
      "18             SNP          ...                      0.0             0.0   \n",
      "19             SNP          ...                      0.0             0.0   \n",
      "20             SNP          ...                      0.0             0.0   \n",
      "21             SNP          ...                      0.0             0.0   \n",
      "22             SNP          ...                      0.0             0.0   \n",
      "23             SNP          ...                      1.0             0.0   \n",
      "24             SNP          ...                      1.0             0.0   \n",
      "25             SNP          ...                      1.0             0.0   \n",
      "26             SNP          ...                      1.0             0.0   \n",
      "27             SNP          ...                      0.0             0.0   \n",
      "28             SNP          ...                      0.0             1.0   \n",
      "29             SNP          ...                      0.0             1.0   \n",
      "...            ...          ...                      ...             ...   \n",
      "44181          SNP          ...                      0.0             0.0   \n",
      "44182          SNP          ...                      0.0             0.0   \n",
      "44183          SNP          ...                      0.0             0.0   \n",
      "44184          SNP          ...                      0.0             0.0   \n",
      "44185          SNP          ...                      0.0             0.0   \n",
      "44186          SNP          ...                      0.0             0.0   \n",
      "44187          SNP          ...                      0.0             0.0   \n",
      "44188          SNP          ...                      0.0             0.0   \n",
      "44189          SNP          ...                      0.0             0.0   \n",
      "44190          SNP          ...                      0.0             0.5   \n",
      "44191          SNP          ...                      0.0             0.5   \n",
      "44192          SNP          ...                      0.0             0.5   \n",
      "44193          SNP          ...                      0.0             0.5   \n",
      "44194          SNP          ...                      0.0             0.5   \n",
      "44195          SNP          ...                      0.0             0.5   \n",
      "44196          SNP          ...                      0.0             1.0   \n",
      "44197          SNP          ...                      0.0             1.0   \n",
      "44198          SNP          ...                      0.0             1.0   \n",
      "44199          SNP          ...                      0.0             1.0   \n",
      "44200          SNP          ...                      0.0             1.0   \n",
      "44201          SNP          ...                      0.0             1.0   \n",
      "44202          SNP          ...                      0.0             1.0   \n",
      "44203          SNP          ...                      0.0             1.0   \n",
      "44204          SNP          ...                      0.0             1.0   \n",
      "44205          SNP          ...                      0.0             1.0   \n",
      "44206          SNP          ...                      0.0             1.0   \n",
      "44207          SNP          ...                      0.0             1.0   \n",
      "44208          SNP          ...                      0.0             0.0   \n",
      "44209          SNP          ...                      0.0             0.0   \n",
      "44210          SNP          ...                      1.0             0.0   \n",
      "\n",
      "      between_0.4_0.5 between_0.5_0.6  distance_prev  distance_next  \\\n",
      "0                 0.0             0.0          55977        55977.0   \n",
      "1                 0.0             0.0          64814        64814.0   \n",
      "2                 0.0             0.0          78214        78214.0   \n",
      "3                 0.0             0.0          87258        87258.0   \n",
      "4                 0.0             0.0          87322        87322.0   \n",
      "5                 0.0             0.0          88779        88779.0   \n",
      "6                 0.0             0.0          89238        89238.0   \n",
      "7                 0.0             0.0          92277        92277.0   \n",
      "8                 0.0             0.0          96413         3138.0   \n",
      "9                 1.0             0.0            636          636.0   \n",
      "10                1.0             0.0          48465        48465.0   \n",
      "11                1.0             0.0          69515          221.0   \n",
      "12                0.0             0.0            294          294.0   \n",
      "13                0.0             0.0           1667         1667.0   \n",
      "14                0.0             0.0           2140         2140.0   \n",
      "15                0.0             0.0           2402         2402.0   \n",
      "16                0.0             0.0           3988         3988.0   \n",
      "17                0.0             0.0           4026         3732.0   \n",
      "18                1.0             0.0          31619         2995.0   \n",
      "19                0.5             0.0          83053        83053.0   \n",
      "20                0.5             0.0          93465        32154.0   \n",
      "21                0.0             1.0             36           36.0   \n",
      "22                0.0             1.0             59          398.0   \n",
      "23                0.0             0.0             12           12.0   \n",
      "24                0.0             0.0           1446         1446.0   \n",
      "25                0.0             0.0           5662         5662.0   \n",
      "26                0.0             0.0           6952         4257.0   \n",
      "27                1.0             0.0          11566        52070.0   \n",
      "28                0.0             0.0            790          790.0   \n",
      "29                0.0             0.0           4530         4530.0   \n",
      "...               ...             ...            ...            ...   \n",
      "44181             0.0             1.0           3726         3726.0   \n",
      "44182             0.0             1.0           3780         3780.0   \n",
      "44183             0.0             1.0           7700         7700.0   \n",
      "44184             0.0             1.0          23453        23453.0   \n",
      "44185             0.0             1.0          25982        25982.0   \n",
      "44186             0.0             1.0          26591         2141.0   \n",
      "44187             1.0             0.0           1143         1143.0   \n",
      "44188             1.0             0.0           2423        62323.0   \n",
      "44189             1.0             0.0           1906       118965.0   \n",
      "44190             0.0             0.5          59010        59010.0   \n",
      "44191             0.0             0.5          59117        59117.0   \n",
      "44192             0.0             0.5          59147        59147.0   \n",
      "44193             0.0             0.5         121990       121990.0   \n",
      "44194             0.0             0.5         126056       126056.0   \n",
      "44195             0.0             0.5         138559           30.0   \n",
      "44196             0.0             0.0         112980       112980.0   \n",
      "44197             0.0             0.0         113031       113031.0   \n",
      "44198             0.0             0.0         114617       114617.0   \n",
      "44199             0.0             0.0         137299       137299.0   \n",
      "44200             0.0             0.0         137690       137690.0   \n",
      "44201             0.0             0.0         139258       139258.0   \n",
      "44202             0.0             0.0         165205       165205.0   \n",
      "44203             0.0             0.0         174256       174256.0   \n",
      "44204             0.0             0.0         177511       177511.0   \n",
      "44205             0.0             0.0         273874       273874.0   \n",
      "44206             0.0             0.0         297423       297423.0   \n",
      "44207             0.0             0.0         424636      9986280.0   \n",
      "44208             0.0             0.0        6287069      6287069.0   \n",
      "44209             0.0             0.0        6430948      5563008.0   \n",
      "44210             0.0             0.0         594887     36391137.0   \n",
      "\n",
      "       Chromosome_y  distance_to_bait_y gc_content_y  SNP_genomic_position  \n",
      "0                 1               51632     0.520000                809687  \n",
      "1                 1               51632     0.520000                809687  \n",
      "2                 1               51632     0.520000                809687  \n",
      "3                 1               51632     0.520000                809687  \n",
      "4                 1               51632     0.520000                809687  \n",
      "5                 1               51632     0.520000                809687  \n",
      "6                 1               51632     0.520000                809687  \n",
      "7                 1               51632     0.520000                809687  \n",
      "8                 1               51632     0.520000                809687  \n",
      "9                 1                  28     0.666667                909238  \n",
      "10                1                  28     0.666667                909238  \n",
      "11                1                  28     0.666667                909238  \n",
      "12                1                  59     0.693333                978974  \n",
      "13                1                  59     0.693333                978974  \n",
      "14                1                  59     0.693333                978974  \n",
      "15                1                  59     0.693333                978974  \n",
      "16                1                  59     0.693333                978974  \n",
      "17                1                  59     0.693333                978974  \n",
      "18                1                  19     0.693333                986732  \n",
      "19                1                  46     0.613333               1021346  \n",
      "20                1                  46     0.613333               1021346  \n",
      "21                1                  33     0.666667               1146965  \n",
      "22                1                  33     0.666667               1146965  \n",
      "23                1                  98     0.746667               1147422  \n",
      "24                1                  98     0.746667               1147422  \n",
      "25                1                  98     0.746667               1147422  \n",
      "26                1                  98     0.746667               1147422  \n",
      "27                1                  10     0.613333               1158631  \n",
      "28                1                  90     0.706667               1222267  \n",
      "29                1                  90     0.706667               1222267  \n",
      "...             ...                 ...          ...                   ...  \n",
      "44181            23                  36     0.586667            3034700462  \n",
      "44182            23                  36     0.586667            3034700462  \n",
      "44183            23                  36     0.586667            3034700462  \n",
      "44184            23                  36     0.586667            3034700462  \n",
      "44185            23                  36     0.586667            3034700462  \n",
      "44186            23                  36     0.586667            3034700462  \n",
      "44187            23                  23     0.640000            3034729194  \n",
      "44188            23                  23     0.640000            3034729194  \n",
      "44189            23                  25     0.613333            3034793940  \n",
      "44190            23                   7     0.613333            3034914811  \n",
      "44191            23                   7     0.613333            3034914811  \n",
      "44192            23                   7     0.613333            3034914811  \n",
      "44193            23                   7     0.613333            3034914811  \n",
      "44194            23                   7     0.613333            3034914811  \n",
      "44195            23                   7     0.613333            3034914811  \n",
      "44196            23                  10     0.480000            3035053400  \n",
      "44197            23                  10     0.480000            3035053400  \n",
      "44198            23                  10     0.480000            3035053400  \n",
      "44199            23                  10     0.480000            3035053400  \n",
      "44200            23                  10     0.480000            3035053400  \n",
      "44201            23                  10     0.480000            3035053400  \n",
      "44202            23                  10     0.480000            3035053400  \n",
      "44203            23                  10     0.480000            3035053400  \n",
      "44204            23                  10     0.480000            3035053400  \n",
      "44205            23                  10     0.480000            3035053400  \n",
      "44206            23                  10     0.480000            3035053400  \n",
      "44207            23                  10     0.480000            3035053400  \n",
      "44208            24              144139     0.386667            3045464316  \n",
      "44209            24              144139     0.386667            3045464316  \n",
      "44210            24              218920     0.600000            3057458272  \n",
      "\n",
      "[44211 rows x 64 columns]\n",
      "44211\n",
      "Index(['Hugo_Symbol_x', 'Entrez_Gene_Id_x', 'Center_x', 'NCBI_Build_x',\n",
      "       'Chromosome', 'Start_position_x', 'End_position_x', 'Strand_x',\n",
      "       'Variant_Classification_x', 'Variant_Type_x',\n",
      "       ...\n",
      "       'between_0.2_0.3', 'between_0.3_0.4', 'between_0.4_0.5',\n",
      "       'between_0.5_0.6', 'distance_prev', 'distance_next', 'Chromosome_y',\n",
      "       'distance_to_bait_y', 'gc_content_y', 'SNP_genomic_position'],\n",
      "      dtype='object', length=103)\n"
     ]
    }
   ],
   "source": [
    "germline_file=\"call_stat_training/small_case_shuffled_medians_segdup_CNV_proxy_map_dist_bait_gc.txt\"\n",
    "somatic_file=\"test_gc.txt\"\n",
    "\n",
    "germline_muts=pd.read_csv(germline_file,sep='\\t')\n",
    "somatic_muts=pd.read_csv(somatic_file,sep='\\t')\n",
    "\n",
    "chr_lens = [0, 249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,\n",
    "              146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,\n",
    "              102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,\n",
    "           155270560]\n",
    "\n",
    "offsets = np.cumsum(chr_lens)\n",
    "\n",
    "germ_chrom=[]\n",
    "som_chrom=[]\n",
    "\n",
    "#make sure all contigs are numerical\n",
    "for idx,row in germline_muts.iterrows():\n",
    "    germ_chrom.append(chrom2int(row.contig))\n",
    "    \n",
    "for idx,row in somatic_muts.iterrows():\n",
    "    som_chrom.append(chrom2int(row.Chromosome))\n",
    "\n",
    "#calculate genomic position\n",
    "germline_muts[\"Chromosome\"]=germ_chrom\n",
    "somatic_muts['numbered_Chromosome']=som_chrom\n",
    "germline_muts[\"genomic_position\"]=pd.to_numeric(germline_muts[\"position\"])+offsets[\n",
    "    pd.to_numeric(germline_muts[\"Chromosome\"])-1\n",
    "]\n",
    "                                         \n",
    "somatic_muts[\"genomic_position\"]=pd.to_numeric(somatic_muts[\"Start_position\"])+offsets[\n",
    "    pd.to_numeric(somatic_muts[\"numbered_Chromosome\"])-1\n",
    "]\n",
    "\n",
    "#sort and combine\n",
    "germline_muts=germline_muts.sort_values(by='genomic_position')\n",
    "germline_muts=germline_muts.reset_index(drop=True)\n",
    "\n",
    "somatic_muts=somatic_muts.sort_values(by='genomic_position')\n",
    "somatic_muts=somatic_muts.reset_index(drop=True)\n",
    "\n",
    "somatic_unique=somatic_muts.drop_duplicates(subset=[\"genomic_position\"])\n",
    "\n",
    "merge_muts=pd.merge_asof(somatic_unique,germline_muts,on='genomic_position',allow_exact_matches=True)\n",
    "merge_muts['SNP_genomic_position']=pd.to_numeric(merge_muts[\"position\"])+offsets[\n",
    "    pd.to_numeric(merge_muts[\"Chromosome_y\"])-1\n",
    "]\n",
    "\n",
    "#get distance to the nearest SNPs\n",
    "merge_muts['distance_prev']=merge_muts['genomic_position']-merge_muts['SNP_genomic_position']\n",
    "\n",
    "merge_muts['distance_next']=abs(merge_muts['genomic_position']-merge_muts['SNP_genomic_position'].shift(-1))\n",
    "\n",
    "#If there is no SNP at a distance greater than the last somatic mutation, calculate distance to the end of the genome\n",
    "merge_muts.loc[len(merge_muts.index)-1,'distance_next']=3088286401-merge_muts.loc[len(merge_muts.index)-2,'genomic_position']\n",
    "\n",
    "#failsafe to check if any of the merges were performed incorrectly\n",
    "negative=np.where(merge_muts['distance_prev']<0)[0]\n",
    "print(negative)\n",
    "\n",
    "somatic_merge=somatic_muts.merge(merge_muts,on=\"genomic_position\")\n",
    "\n",
    "somatic_merge.to_csv(\"test_prepared.txt\",sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
